File: DTFlinkBatch/src/main/java/com/kevin/java/sink/CsvDemo.java
package com.kevin.java.sink;

import org.apache.flink.api.common.functions.FlatMapFunction;
import org.apache.flink.api.java.ExecutionEnvironment;
import org.apache.flink.api.java.operators.AggregateOperator;
import org.apache.flink.api.java.operators.DataSource;
import org.apache.flink.api.java.tuple.Tuple2;
import org.apache.flink.util.Collector;

/**
 * @author caonanqing
 * @version 1.0
 * @description
 * @createDate 2020/4/12
 */
public class CsvDemo {

    public static void main(String[] args) throws Exception {
        String inputPath = "DTFlinkBatch/src/main/resources/words.txt";
        String outPath = "D:/big_data/result.csv";

        // 获取运行环境
        ExecutionEnvironment env = ExecutionEnvironment.getExecutionEnvironment();
        // 获取文件中的内容
        DataSource<String> text = env.readTextFile(inputPath);
        // 切分数据，初始化单词数值为1，按照第一位元素进行分组，第二位元素进行相加
        AggregateOperator<Tuple2<String, Integer>> counts = text.flatMap(new FlatMapFunction<String, Tuple2<String, Integer>>() {
            @Override
            public void flatMap(String s, Collector<Tuple2<String, Integer>> out) throws Exception {
                for (String value : s.split(" ")) {
                    out.collect(new Tuple2<String, Integer>(value, 1));
                }
            }
        }).groupBy(0).sum(1);

        counts.writeAsCsv(outPath,"\n"," ").setParallelism(1);
        env.execute("csv demo");

    }
}

File: dt-flink-dataset/src/main/scala/com/kevin/scala/sink/CsvDemo.scala
package com.kevin.scala.sink

import org.apache.flink.api.scala.{DataSet, ExecutionEnvironment}

/**
 * @author caonanqing
 * @version 1.0
 * @description   将结果输出到csv文件中
 * @createDate 2020/4/12
 */
object CsvDemo {

  def main(args:Array[String]):Unit = {

    val inputPath = "DTFlinkBatch\\src\\main\\resources\\words.txt"
    val outPath = "D:\\big_data\\result1"
    // 获取运行环境
    val env:ExecutionEnvironment = ExecutionEnvironment.getExecutionEnvironment

    val text = env.readTextFile(inputPath)
    // 引用隐式转换
    import org.apache.flink.api.scala._
    // 转换成小写，并切分，过滤掉为空值，将单词数值初始化为1，第一个元素作为分组，第二个元素进行求和相加
    val counts = text.flatMap(_.toLowerCase.split(" "))
      .filter(_.nonEmpty)
      .map((_,1))
      .groupBy(0)
      .sum(1)
    // 将结果数据保存到csv中，每行以\n切分，每列以空格切分，设置分区为2个（输出到两个文件）
    counts.writeAsCsv(outPath,"\n"," ").setParallelism(2)
    // 执行程序
    env.execute("csv demo")

  }

}
